File: 프로그래머스/3/42895. N으로 표현/N으로 표현.js
function solution(N, number) {
  if (N === number) return 1;

  const dp = Array.from({ length: 9 }, () => new Set()); // dp[i] = N을 i번 써서 만들 수 있는 값들

  for (let i = 1; i <= 8; i++) {
    // 이어붙인 수: N, NN, NNN...
    const concat = Number(String(N).repeat(i));
    dp[i].add(concat);

    // j + (i-j)로 분할해서 조합
    for (let j = 1; j < i; j++) {
      for (const a of dp[j]) {
        for (const b of dp[i - j]) {
          dp[i].add(a + b);
          dp[i].add(a - b);
          dp[i].add(a * b);
          if (b !== 0) dp[i].add(Math.trunc(a / b)); // 나머지 무시(0쪽으로 절삭)
        }
      }
    }

    if (dp[i].has(number)) return i;
  }

  return -1;
}

File: 프로그래머스/3/42895. N으로 표현/N으로 표현.py
def solution(N, number):
    if N == number:
        return 1

    dp = [set() for _ in range(9)]  # 0은 안 쓰고 1~8 사용

    for i in range(1, 9):
        # 이어붙인 숫자 (ex: 5, 55, 555 ...)
        concat = int(str(N) * i)
        dp[i].add(concat)

        # j + (i-j)로 쪼개서 사칙연산 조합
        for j in range(1, i):
            for a in dp[j]:
                for b in dp[i - j]:
                    dp[i].add(a + b)
                    dp[i].add(a - b)
                    dp[i].add(a * b)
                    if b != 0:
                        dp[i].add(a // b)  # 나머지 버림

        if number in dp[i]:
            return i

    return -1